tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c

```diff
@@ -0,0 +1,58 @@
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define MAP_SIZE 4096UL
+#define MAP_MASK (MAP_SIZE - 1)
+
+int main(int argc, char** argv)
+{
+    int fd;
+    void *map_base, *virt_addr;
+    uint32_t read_result, write_result;
+
+    uint32_t writeval = 't';
+    off_t address = 0x00000001;
+    int retval = -1;
+
+    /* Open character device */
+    if ((fd = open("/dev/mem", O_RDWR)) == -1) {
+        perror("cannot open /dev/mem");
+        goto close;
+    }
+
+    /* Map one page */
+    map_base = mmap(NULL, MAP_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, address & ~MAP_MASK);
+    if (map_base == MAP_FAILED) {
+        perror("cannot map physical memory");
+        goto unmap;
+    }
+    printf("Memory mapped at address %p.\n", map_base);
+
+    /* Read memory map */
+    virt_addr = (char*)map_base + (address & MAP_MASK);
```

Comment:
IIRC busybox has a `devmem` applet that does something similar. Maybe you could use it in the snap.